s3/src/main/java/com/yahoo/ycsb/db/S3Client.java

```diff
@@ -0,0 +1,348 @@
+  * @return Zero on success, a non-zero error code on error. See this class's
+  *         description for a discussion of error codes.
+  */
+  @Override
+  public int update(String bucket, String key, 
+        HashMap<String, ByteIterator> values) {
+    return writeToStorage(bucket, key, values, 1);
+  }
+  /**
+  * Perform a range scan for a set of files in the bucket. Each
+  * field/value pair from the result will be stored in a HashMap.
+  * 
+  * @param bucket
+  *            The name of the bucket
+  * @param startkey
+  *            The file key of the first file to read.
+  * @param recordcount
+  *            The number of files to read
+  * @param fields
+  *            The list of fields to read, or null for all of them
+  * @param result
+  *            A Vector of HashMaps, where each HashMap is a set field/value
+  *            pairs for one file
+  * @return Zero on success, a non-zero error code on error. See the
+  *         {@link DB} class's description for a discussion of error codes.
+  */
+  @Override
+  public int scan(String bucket, String startkey, int recordcount, 
+        Set<String> fields, Vector<HashMap<String, ByteIterator>> result) {
+    return scanFromStorage(bucket, startkey, recordcount, result);
+  }
+
+  protected int writeToStorage(String bucket, String key, 
+        HashMap<String, ByteIterator> values, int updateMarker) {
+    int totalSize = 0;
+    int fieldCount = values.size(); //number of fields to concatenate
+    // getting the first field in the values
+    Object keyToSearch = values.keySet().toArray()[0]; 
+    // getting the content of just one field
+    byte[] sourceArray = values.get(keyToSearch).toArray(); 
+    int sizeArray = sourceArray.length; //size of each array
+    if (updateMarker == 0){
+      totalSize = sizeArray*fieldCount;
+    } else {
+      try {
+        S3Object object = 
+            this.s3Client.getObject(new GetObjectRequest(bucket, key));
+        ObjectMetadata objectMetadata = 
+            this.s3Client.getObjectMetadata(bucket, key);
+        int sizeOfFile = (int)objectMetadata.getContentLength();
+        fieldCount = sizeOfFile/sizeArray;
+        totalSize = sizeOfFile;
+      } catch (Exception e){
+        e.printStackTrace();
+        return 1;
+      }
+    }
+    byte[] destinationArray = new byte[totalSize];
+    int offset = 0;
+    for (int i = 0; i < fieldCount; i++) {
+      System.arraycopy(sourceArray, 0, destinationArray, offset, sizeArray);
+      offset += sizeArray;
+    }
+    InputStream input = new ByteArrayInputStream(destinationArray);
+    ObjectMetadata metadata = new ObjectMetadata();
+    metadata.setContentLength(totalSize);
+    try {
+      PutObjectResult res = 
+          this.s3Client.putObject(bucket, key, input, metadata);
+      if(res.getETag() == null) {
+        return 1;
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+      return 1;
+    } finally {
+      try {
+        input.close();
```

Comment:
ByteArrayInputStream doesn't do anything in close. If you're following good practice for input being typed as InputStream, switch to try-with-resources instead.